zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+            # until we reach a leaf
+            subtree_root = root.right_child
+            while subtree_root.left_child:
+                peaks.push(tmp_root.left_child)
+                subtree_root = tmp_root.right_child
+
+        new_root = bag_peaks(peaks)
+        return new_root
+
+    def bag_peaks(peaks: List[ZcashMMRNode]) -> ZcashMMRNode:
+        '''
+        "Bag" a list of peaks, and return the final root
+        '''
+        root = peaks[0]
+        for i in range(1, len(peaks)):
+            root = make_parent(root, peaks[i])
+        return root
+
+Header modifications specification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ZIP introduces a new header version. It is identical to the current v4 header [#zcashBlock]_, except for the following changes:
+
+1. The version number is changed to `5`.
+2. ``hashSaplingFinalRoot`` is replaced by ``hashChainHistoryRoot``, as described above.
+
+The new block header format is:
+
+.. code-block:: C
+
+    class CBlockHeader
+    {
+    public:
+        // header
+        static const size_t HEADER_SIZE=4+32+32+32+4+4+32; // excluding Equihash solution
+        static const int32_t CURRENT_VERSION=5;
+        int32_t nVersion;
+        uint256 hashPrevBlock;
+        uint256 hashMerkleRoot;
+        uint256 hashChainHistoryRoot;
+        uint32_t nTime;
+        uint32_t nBits;
+        uint256 nNonce;
+        std::vector<unsigned char> nSolution;
+        ...
+    }
+
+
+Rationale
+-----------
+
+Tree nodes
+~~~~~~~~~~~
+
+Nodes in the commitment tree are canonical and immutable. They are cheap to generate, as (with the exception of ``nShieldedTxCount``) all metadata is already generated during block construction and/or checked during block validation. Nodes are relatively compact in memory. Approximately 140,000 blocks have elapsed since Sapling activation. Assuming a 164 byte commitment to each of these, we would have generated approximately 24 MB of additional storage cost for the set of leaf nodes (and an additional ~24 MB for storage of intermediate nodes).
+
+``hashSubtreeCommitment`` forms the strucuture of the commitment tree. Other metadata commitments were chosen to serve specific purposes. Variable-length commitments are placed last, so that most metadata in a node can be directly indexed. We considered using fixed-length commitments here, but opted for variable-length, in order to marginally reduce the memory requirements for managing and updating the commitment trees.
+
+In leaf nodes, some information is repeated. We chose to do this so that leaf nodes could be treated identically to internal and root nodes for all algorithms and (de)serializers. Leaf nodes are easily identifiable, as they will show proof of work in the ``hashSubtreeCommitment`` field, and their block range (calculated as ``nLatestHeight - nEarliestHeight + 1``) will be precisely 1. For the same reason, we change the semantics of ``hashSubtreeCommitment`` in leaf nodes to commit
```

Comment:
This is slightly clearer as to what is being calculated:

```suggestion
In leaf nodes, some information is repeated. We chose to do this so that leaf nodes could be treated identically to internal and root nodes for all algorithms and (de)serializers. Leaf nodes are easily identifiable, as they will show proof of work in the ``hashSubtreeCommitment`` field, and their block range (calculated as ``nLatestHeight - (nEarliestHeight - 1)``) will be precisely 1. For the same reason, we change the semantics of ``hashSubtreeCommitment`` in leaf nodes to commit
```

Reply:
The last line of this paragraph is unfinished.